c_src/bayesian.c

```diff
@@ -0,0 +1,208 @@
+        const double sum = p1->y + p2->y;
+        const double p1_prob = p1->y / sum;
+
+        for (size_t j=0; j<dims; ++j) {
+            p->x[j] = (random_f64() < p1_prob) ? p1->x[j] : p2->x[j];
+        }
+
+        for (size_t j=0; j<dims; ++j) {
+            if (random_f64() < MUTATE_PROB) {
+                p->x[j] = lower[j] + (upper[j] - lower[j]) * random_f64();
+            }
+        }
+
+        double variance;
+        double y = randomforest_full_predict(forest, p->x, &variance);
+        p->y = ei(y, sqrt(variance), fmin);
+
+        if (p->y > best) {
+            best = p->y;
+            for (size_t j=0; j<dims; ++j) { out[j] = p->x[j]; }
+        }
+
+        DataPoint* temp = points[oldest];
+        points[oldest] = p;
+        p = temp;
+
+        oldest = (oldest + 1) % EVO_POPULATION;
+
+    }
+
+    // Make sure to update the backing pointer array, since p may have been swapped
+    points[EVO_POPULATION] = p;
+}
+
+void shuffle_points(DataPoint** points, const size_t n) {
+    for (size_t i=0; i<n-1; ++i) {
+        const size_t j = i + random_u64() % (n-i);
+        DataPoint* temp = points[i];
+        points[i] = points[j];
+        points[j] = temp;
+    }
+}
+
+double bayesian_optimization(OptFunc f, const double* lower, const double* upper, double* x, const size_t dims,
+                             const size_t doe, const size_t iter) {
+
+    const size_t dp_size = sizeof(DataPoint) + dims * sizeof(double);
+
+    // Alloc backing data for all points (1 per iteration)
+    DataPoint* data = calloc(iter, dp_size);
+    
+    // Alloc pointers that can point into backing data points
+    DataPoint** points = calloc(iter, sizeof(DataPoint*));
+
+    // Assign each pointer to an allocated data point
+    points[0] = data;
+    for (size_t i=1; i<iter; ++i) { points[i] = (void*)points[i-1] + dp_size; }
+
+    // Alloc work space for local search algorithm
+    DataPoint* workspace = calloc(EVO_POPULATION + 1, dp_size);
+
+    // Pointers into workspace
+    DataPoint** workspace_points = calloc(EVO_POPULATION + 1, sizeof(DataPoint*));
+
+    workspace_points[0] = workspace;
+    for (size_t i=1; i<EVO_POPULATION+1; ++i) {
+        workspace_points[i] = (void*)workspace_points[i-1] + dp_size;
+    }
+
+    // This will correspond to the point in the parameter x
+    double best_val = INFINITY;
+
+    // Initialize DoE phase as a latin hypercube
+    for (size_t i=0; i<dims; ++i) {
+        for (size_t j=0; j<doe; ++j) {
+            points[j]->x[i] = lower[i] + (upper[i] - lower[i]) * (j + random_f64()) / doe;
+        }
+        shuffle_points(points, doe);
+    }
+
+    // Evaluate DoE points
+    for (size_t i=0; i<doe; ++i) {
+        const double y = f(points[i]->x);
+        points[i]->y = y;
+
+        if (y < best_val) {
+            best_val = y;
+            for (size_t j=0; j<dims; ++j) { x[j] = points[i]->x[j]; }
+        }
+    }
+
+    double mean_iter = 0.0;
+
+    for (size_t i=doe; i<iter; ++i) {
```

Comment:
One little detail here. In the python version we count the DoE and BO iterations separately. I.e. we start the loop with i=0

Reply:
I can update it to match the rest of HM